cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/worker/WorkerProgramRunner.java

```diff
@@ -110,6 +120,23 @@ public ProgramController run(Program program, ProgramOptions options) {
     return controller;
   }
 
+  @Nullable
+  private AdapterSpecification getAdapterSpecification(Arguments arguments) {
```

Comment:
Thinking out loud - These methods would potentially be required in every program that is supported through app templates, right? with very similar logic? Is there a way to abstract them?

Perhaps static methods ``AdapterSpecification.from(Arguments)`` (or better ``AdapterSpecification.from(@Nullable String json)``, ``PluginInstantiator.from(@Nullable AdapterSpecification, ClassLoader)``?